### ISSUE TITLE
Changing the seed for VideoReader appears to do nothing

### ISSUE THREAD

**Issue by mark-aai** [NVIDIA/DALI#1695]
Hi,

I've found that when using random_shuffle=TRUE, that regardless of what seed I pass to either Pipeline or VideoPipeline, it always creates the same order of shuffled data. This is most easily checked with initial_fill=1.

This is my relevant code and sample output:
`  
    batch_size=5
    sequence_length=7
    device_id=0
    initial_fill=1
    num_classes = 20
    datafile = 'data.txt'

    class VideoPipeline(Pipeline):
        def __init__(self, batch_size, initial_fill=1,device_id=0, num_threads=4, seed1=0, seed2=1,shuffle=True, data_file=None, sequence_length=1,stride=1):

            super(VideoPipeline, self).__init__(batch_size, num_threads, device_id, seed1)
            self.reader = ops.VideoReader(device='gpu',  #VideoReader does not seem to have cpu support
                                          file_list=data_file, 
                                          sequence_length=sequence_length,
                                          shard_id=0,
                                          num_shards=1,
                                          enable_frame_num=True,
                                          file_list_frame_num=True,
                                          random_shuffle=shuffle,
                                          initial_fill=initial_fill,
                                          stride=stride,
                                          dtype=types.UINT8,
                                          seed=seed2
                                         )

        def define_graph(self):
            data, labels, frame_num = self.reader(name="Reader")

            return data, labels, frame_num

    seed1=41
    seed2=61
    pipe1 = VideoPipeline(batch_size=batch_size, 
                         device_id = device_id,
                         num_threads=2,
                         seed1=seed1,
                         seed2=seed2,
                         shuffle=True,
                         data_file=datafile,
                         sequence_length=sequence_length,
                         stride = 3,
                         initial_fill=initial_fill
                        )
    pipe1.build()    

    raw_data, labels, frame_num=pipe1.run()
    frame_num = frame_num.as_cpu().as_array()
    labels = labels.as_cpu().as_array()
    print('seed1=',seed1,'; seed2=',seed2)
    print(frame_num,labels)

    seed1=37
    seed2=67
    pipe2 = VideoPipeline(batch_size=batch_size, 
                         device_id = device_id,
                         num_threads=2,
                         seed1=seed1,
                         seed2=seed2,
                         shuffle=True,
                         data_file=datafile,
                         sequence_length=sequence_length,
                         stride = 3,
                         initial_fill=initial_fill
                        )
    pipe2.build()    

    raw_data, labels, frame_num=pipe2.run()
    frame_num = frame_num.as_cpu().as_array()
    labels = labels.as_cpu().as_array()
    print('seed1=',seed1,'; seed2=',seed2)
    print(frame_num,labels)`

> 
seed1= 41 ; seed2= 61
[[  169]
 [11683]
 [14533]
 [16183]
 [ 4088]] [[100002020]
 [ 42003080]
 [ 45002130]
 [ 44002091]
 [ 77002091]]
seed1= 37 ; seed2= 67
[[  169]
 [11683]
 [14533]
 [16183]
 [ 4088]] [[100002020]
 [ 42003080]
 [ 45002130]
 [ 44002091]
 [ 77002091]]

Am I doing something wrong? My expectation was that changing the seed would change the randomly chosen order in which data is shuffled.

**Comment by JanuszL** [comment]
Hi,
According to `initial_fill` documentation - ` Size of the buffer used for shuffling. If random_shuffle is off then this parameter is ignored.`. So if you set to 1 then you are asking to randomly sample buffer of size 1 what defeats the idea of shuffling. If you increase it to 2 then it should work as executed (still entropy won't be that good).
DALI reader traverses the data set in the same order in each epoch filing the internal shuffling buffer which is randomly sampled to create a batch. The smaller the buffer is the less random the shuffling becomes. In the case of Video, it consumes a lot of memory, still, it should be bigger than 1.

**Comment by mark-aai** [comment]
Thankyou for the quick reply. If random_shuffle only changes the initial_fill length, then what you say makes sense, and a buffer size of 1 will do nothing.

However, I need to clarify that this is independent of the  undesirable behavior I have observed with random_shuffle.

What I may not have made clear is that with random_shuffle set to True, data is returned in a complete random ordering relative to the ordering of the rows of the data file, independently of initial_fill. This is what I would expect. Additionally, if I the increase initial_fill size from, then both things occur: random shuffling of the entire file, which I infer happensfirst, plus random shuffling across the buffer size.

The problem is that changing the random seed in any random number generator would be expected to change the random ordering in which rows are returned, but it does not.  Any random number generator where a seed is passed should produce entirely different results for each seed. But here, it does not.

**Comment by JanuszL** [comment]
> What I may not have made clear is that with random_shuffle set to True, data is returned in a complete random ordering relative to the ordering of the rows of the data file, independently of initial_fill.

DALI does global preshuffling to avoid having all entries with class 0, then class 1, 2, 3.... when reading. Using sampling buffer on its own is not sufficient in that case as if it is smaller than a number of samples in the given class then you would get random samples but inside the same class. This preshuffling is deterministic as you want to have the same sharing in each shard, even in different nodes, so this is intended.

**Comment by mark-aai** [comment]
Thanks, ok I understand. But in my experience best practice for neural network training requires a new random ordering for each epoch, and the sampling buffer cannot achieve this unless its size is the same as the total number of samples.  

It might be of value to others to highlight that recreating a data file with a new ordering certainly works for me using a single shard, and has the added advantage that because I designate start and end frames in short sequences of about 20 frames, I can jitter the start and end frames randomly in each creation of the data file for each epoch, for temporal data augmentation. I couldn't work out a way to do that using Dali operations, and its also more efficient then loading extra frames and then dropping some in the cpu data generator. 

Despite these workarounds, using Dali to load directly from MP4 files is massively better than what I was doing before, i.e. saving all frames as images. That method is quickly limited by hard drive space. So thanks and I hope to keep seeing upgrades of video ops, such as resize in the near future.

**Comment by JanuszL** [comment]
> Thanks, ok I understand. But in my experience best practice for neural network training requires a new random ordering for each epoch, and the sampling buffer cannot achieve this unless its size is the same as the total number of samples.

In general yes, but not always. There are some containers - like TFRecord, LMDB or RecordIO which doesn't provide fast random access and data incise is already preshuffled. FWs - like TensorFlow reads this data sequentially and use internal sampling buffer - exactly what we do.
